# Introduction

« *Un réseau de neurones artificiels est un modèle de calcul dont la conception est très schématiquement inspirée du fonctionnement des neurones biologiques.* »

Source : https://fr.wikipedia.org/wiki/Réseau_de_neurones_artificiels 

Dans ce TP, nous allons illustrer le fonctionnement d'un réseau de neurones destiné à la reconnaissance de caractères (ici des chiffres de 0 à 9).

# Fonctionnement

<img src="https://github.com/PlantecJY/python-mic/raw/master/rdn.png" alt="Drawing" style="width: 200px;"/>

La figure ci-dessus schématise un réseau de neurones artificiel. Il est constitué de neurones reliés entre eux et figurés par des cercles de couleur.

Un neurone de base (cercle bleu) possède plusieurs entrées par lesquelles il reçoit un signal issu d'autres neurones et plusieurs sorties par lesquelles il peut sous certaines conditions émettre un signal vers d'autres neurones. 

Certains neurones – les neurones d'entrée (cercles verts) – sont activés en recevant une information directement depuis l'extérieur. Les neurones de sortie (ici, un seul cercle jaune), quant à eux, ne sont pas nécessairement du même nombre que les neurones d'entrée ; on s'intéresse seulement à leur état, car il ne retransmettent pas d'information.

Le neurone est conçu comme un automate doté d'une fonction de transfert qui transforme ses entrées notéesxien une sortie selon des règles précises. 

Dans le modèle élémentaire que nous allons mettre en œuvre, l'information qui parcourt une entrée vaut 0 (pas de signal) ou 1 (un signal). Le neurone effectue une somme pondérée – on parle de « poids synaptique » – de chacune de ses entrées (en effet, l'efficacité de la transmission des signaux d'un neurone à l'autre peut varier), compare la somme résultante à une valeur seuil, et si cette somme est supérieure ou égale à ce seuil, il s'active et émet un signal (modèle ultra-simplifié du fonctionnement d'un neurone biologique) ; dans le cas contraire, il reste inactif.

Le seuil de chaque neurone ainsi que la valeur des poids synaptiques wij sont donc des paramètres importants du réseau : selon les informations d'entrée (dans ce TP, un pattern de 0 et de 1), les neurones de sortie seront activés ou pas. Le réseau est donc un outil permettant d'opérer une classification des patterns qui lui sont soumis ; notons que plusieurs patterns d'entrée différents peuvent conduire au même pattern de sortie.

L'idée est de choisir ces poids et ce seuil de telle façon que la classification opérée par le réseau soit aussi proche que possible de celle voulue par le concepteur. C’est ce qu'on nomme la phase d’apprentissage du réseau. Pour un réseau de neurones formels, apprendre revient donc à déterminer le jeu de coefficients synaptiques permettant de classifier les données présentées ; lorsque ce jeu est déterminé, il est attendu du réseau qu'il classifie correctement une nouvelle entrée.

Dans un modèle plus riche, le neurone fonctionne avec des nombres réels (souvent compris dans l’intervalle [0,1] ou [-1,1]).

# Exemple choisi

Nous vous proposons d'étudier un réseau de neurones destiné à la reconnaissance de chiffres.
Le réseau possède 30 neurones d'entrée auxquels sont soumis des patterns de 30 informations binaires correspondant aux pixels parcourus de gauche à droite, ligne par ligne. Dans la figure ci-dessous, c'est le pattern correspondant au chiffre 1 qui est envoyé.

<img src="https://github.com/PlantecJY/python-mic/raw/master/principe.png" alt="Drawing" style="width: 200px;"/>

 → 111110000100001000010000100001

Le réseau possède 10 neurones de sortie correspondant chacun à un chiffre de 0 à 9. En sortie, seul le neurone correspondant au chiffre 1 est actif.

Le réseau pourra avoir le même comportement pour un pattern d'entrée bruité, mais correspondant toujours au chiffre 1.

# Travail à effectuer

Le programme fait l’objet du présent notebook. Il est découpé en une série de fonctions. Ce découpage vous permet d'avancer progressivement en testant chaque fonction. La réalisation du programme s'appuie sur les résultats des précédentes séances.

- Structuration des données pour l’apprentissage
- Création du réseau de neurones
- Test du réseau de neurones
- Apprentissage
- Test du réseau de neurones

Nota : lors de l'examen final, plusieurs questions porteront sur ce projet.

# Réalisation

## Structuration des données pour l’apprentissage
Cette étape s'appuie principalement sur le travail effectué lors de la séance numéro 1. Avant tout, assurez-vous d'avoir à disposition votre réponse à la dernière question de cette séance.

**Exercice** : Créer une fonction nommée *structuration_donnees* qui permet :
- de récupérer des données correspondant aux differentes facons de representer les chiffres ;
- d'applatir ces matrices sous la forme de listes ;
- d'orgasnier ces listes sous la forme d’un dictionnaire.


In [1]:
nb0a = [[1,1,1,1,1],
        [1,0,0,0,1],
        [1,0,0,0,1],
        [1,0,0,0,1],
        [1,0,0,0,1],
        [1,1,1,1,1]]

nb0b = [[0,1,1,1,0],
        [1,0,0,0,1],
        [1,0,0,0,1],
        [1,0,0,0,1],
        [1,0,0,0,1],
        [0,1,1,1,0]]

nb1a = [[0,0,1,0,0],
        [0,0,1,0,0],
        [0,0,1,0,0],
        [0,0,1,0,0],
        [0,0,1,0,0],
        [0,0,1,0,0]]

nb1b = [[0,0,0,0,0],
        [0,0,1,0,0],
        [0,0,1,0,0],
        [0,0,1,0,0],
        [0,0,1,0,0],
        [0,0,0,0,0]]

nb2a = [[1,1,1,1,1],
        [0,0,0,0,1],
        [1,1,1,1,1],
        [1,0,0,0,0],
        [1,0,0,0,0],
        [1,1,1,1,1]]

nb2b = [[1,1,1,1,1],
        [0,0,0,0,1],
        [0,0,0,0,1],
        [1,1,1,1,1],
        [1,0,0,0,0],
        [1,1,1,1,1]]

nb3a = [[1,1,1,1,1],
        [0,0,0,0,1],
        [0,0,0,0,1],
        [1,1,1,1,1],
        [0,0,0,0,1],
        [1,1,1,1,1]]

nb3b = [[1,1,1,1,1],
        [0,0,0,0,1],
        [1,1,1,1,1],
        [0,0,0,0,1],
        [0,0,0,0,1],
        [1,1,1,1,1]]

nb4a = [[1,0,1,0,0],
        [1,0,1,0,0],
        [1,0,1,0,0],
        [1,1,1,1,1],
        [0,0,1,0,0],
        [0,0,1,0,0]]

nb4b = [[0,0,0,1,0],
        [0,0,1,1,0],
        [0,1,0,1,0],
        [1,1,1,1,1],
        [0,0,0,1,0],
        [0,0,0,1,0]]

nb5a = [[1,1,1,1,1],
        [1,0,0,0,0],
        [1,1,1,1,1],
        [0,0,0,0,1],
        [0,0,0,0,1],
        [1,1,1,1,1]]

nb5b = [[1,1,1,1,1],
        [1,0,0,0,0],
        [1,0,0,0,0],
        [1,1,1,1,1],
        [0,0,0,0,1],
        [1,1,1,1,1]]

nb5c = [[1,1,1,1,1],
        [1,0,0,0,0],
        [1,1,1,1,0],
        [0,0,0,0,1],
        [0,0,0,0,1],
        [1,1,1,1,0]]

nb6a = [[1,1,1,1,1],
        [1,0,0,0,0],
        [1,1,1,1,1],
        [1,0,0,0,1],
        [1,0,0,0,1],
        [1,1,1,1,1]]

nb6b = [[1,1,1,1,1],
        [1,0,0,0,0],
        [1,0,0,0,0],
        [1,1,1,1,1],
        [1,0,0,0,1],
        [1,1,1,1,1]]

nb6c = [[0,1,1,1,1],
        [1,0,0,0,0],
        [1,1,1,1,0],
        [1,0,0,0,1],
        [1,0,0,0,1],
        [0,1,1,1,0]]

nb7a = [[1,1,1,1,1],
        [0,0,0,0,1],
        [0,0,1,1,1],
        [0,0,0,0,1],
        [0,0,0,0,1],
        [0,0,0,0,1]]

nb7b = [[1,1,1,1,1],
        [0,0,0,0,1],
        [0,0,0,1,0],
        [0,0,1,0,0],
        [0,1,0,0,0],
        [1,0,0,0,0]]

nb8a = [[1,1,1,1,1],
        [1,0,0,0,1],
        [1,1,1,1,1],
        [1,0,0,0,1],
        [1,0,0,0,1],
        [1,1,1,1,1]]

nb8b = [[1,1,1,1,1],
        [1,0,0,0,1],
        [1,0,0,0,1],
        [1,1,1,1,1],
        [1,0,0,0,1],
        [1,1,1,1,1]]

nb8c= [[0,1,1,1,0],
        [1,0,0,0,1],
        [1,1,1,1,1],
        [1,0,0,0,1],
        [1,0,0,0,1],
        [0,1,1,1,0]]

nb9a = [[1,1,1,1,1],
        [1,0,0,0,1],
        [1,1,1,1,1],
        [0,0,0,0,1],
        [0,0,0,0,1],
        [1,1,1,1,1]]

nb9b = [[1,1,1,1,1],
        [1,0,0,0,1],
        [1,0,0,0,1],
        [1,1,1,1,1],
        [0,0,0,0,1],
        [1,1,1,1,1]]

nb9c = [[0,1,1,1,0],
        [1,0,0,0,1],
        [0,1,1,1,1],
        [0,0,0,0,1],
        [0,0,0,0,1],
        [1,1,1,1,0]]

def print_diese():
    print("#", end="")

def print_espace():
    print(" ", end="")

def print_saut_de_ligne():
    print("")
    
def flatten_matrice(m):
    # Corps de la fonction a completer.
    ret = []
    for line in m:
        ret += line
    return ret



In [2]:
def structuration_donnees():
    # Algorithme de la fonction :
    # 1. Creation de la variable d correspondant au dictionnaire vide
    # 2. Creation des entrees de ce dictionnaire : chacune des cles correspond a un chiffre
        # Associer a chaque cle une liste vide dans le dictionnaire
    # 3. Pour chaque chiffre, represente par sa matrice :
        # Applatir la matrice avec la fonction flatten_matrice
        # Ajouter la liste obtenue dans la bonne entree du dictionnaire
    # Retour du dictionnaire.
    
    ### START CODE HERE ###
    d = {
            0: [flatten_matrice(nb0a), flatten_matrice(nb0b)],
            1: [flatten_matrice(nb1a), flatten_matrice(nb1b)],
            2: [flatten_matrice(nb2a), flatten_matrice(nb2b)],
            3: [flatten_matrice(nb3a), flatten_matrice(nb3b)],
            4: [flatten_matrice(nb4a), flatten_matrice(nb4b)],
            5: [flatten_matrice(nb5a), flatten_matrice(nb5b), flatten_matrice(nb5c)],
            6: [flatten_matrice(nb6a), flatten_matrice(nb6b), flatten_matrice(nb6c)],
            7: [flatten_matrice(nb7a), flatten_matrice(nb7b)],
            8: [flatten_matrice(nb8a), flatten_matrice(nb8b), flatten_matrice(nb8c)],
            9: [flatten_matrice(nb9a), flatten_matrice(nb9b), flatten_matrice(nb9c)]
    }
    
    ### END CODE HERE ###
    return d

d = structuration_donnees()



**Exercice** : Récupérez la fonction *print_matrice* qui permet d'afficher un chiffre un ascii.

In [3]:
def print_matrice(m):
    ### START CODE HERE ###
    i = 0
    for nb in m:
        if nb == 1:
            print_diese()
        else:
            print_espace()
        i += 1
        if i % 5 == 0:
            print_saut_de_ligne()
            i = 0
    ### END CODE HERE ###

**Test** :

In [4]:
print_matrice(d[9][2])

 ### 
#   #
 ####
    #
    #
#### 


**Sortie attendue** :

     ### 
    #   #
     ####
        #
        #
    ####

## Création du réseau de neurones

Cette étape s'appuie principalement sur le travail effectué lors de la séance numéro 2.
Avant tout, assurez-vous d'avoir à disposition votre réponse à la dernière question de cette séance.

**Exercice** : Créer une fonction nommée *init_poids* qui va créer une matrice qui va contenir les poids du réseau de neurones. Elle accepte deux paramètres qui sont :
- *dimEntree* : le nombre de pixels qui représente le chiffre et
- *dimSortie* : le nombre de sorties du réseau.

Cette fonction doit réaliser les étapes suivantes :
- créer une matrice poids de dimension *dimEntree* x *dimSortie* ;
- initialiser le contenu de cette matrice avec des valeurs aléatoire comprises entre -50 et 50.
- retourner cette matrice.

In [5]:
from numpy import random

def random_within(a, b):
    return (b-a) * random.random() + a

def random_list(n, a, b):
    return [random_within(a, b) for i in range(n)]

def init_poids(dimEntree, dimSortie):
    random.seed(2) # NE PAS TOUCHER CETTE LIGNE QUI REND DETERMINISTE LE RESULTAT
    
    ### START CODE HERE ###
    poids = [random_list(dimSortie, -50, 50) for i in range(dimEntree)]
    ### END CODE HERE ###
    return poids



**Test** :

In [6]:
poids = init_poids(30, 10)
poids[0]

[-6.400509785799628,
 -47.40737681721087,
 4.966247787870913,
 -6.467760738172309,
 -7.963219791251099,
 -16.96651789961259,
 -29.53513659621575,
 11.92709663506637,
 -20.034532632547688,
 -23.317272489713336]

**Sortie attendue** :

      [-6.400509785799628,
       -47.40737681721087,
       4.966247787870913,
       -6.467760738172309,
       -7.963219791251099,
       -16.96651789961259,
       -29.53513659621575,
       11.92709663506637,
       -20.034532632547688,
       -23.317272489713336]

## Test du réseau de neurones

**Exercice** : Créer une fonction nommée *calcul_neurone* qui calcule la sortie d'un seul neurone *j* à partir d'une entrée *e* représentant un chiffre et de la matrice *poids*.

Attention, la sortie pour l'instant n'a aucun lien avec l'entrée car l'apprentissage n'a pas encore eu lieu.

In [7]:
[(i*j) for (i, j) in zip(d[9][0], [l[0] for l in poids])]

[-6.400509785799628,
 12.11338327692949,
 0.5246090121703944,
 -3.2212515417699734,
 46.45510800892552,
 4.420816014810214,
 0.0,
 0.0,
 0.0,
 -18.674147157028642,
 27.80081598084682,
 -23.62214707199802,
 4.790778009037666,
 -40.47119477391963,
 14.331142724823522,
 -0.0,
 0.0,
 0.0,
 -0.0,
 44.04504935420886,
 0.0,
 0.0,
 0.0,
 0.0,
 -11.920741628379197,
 -37.9960207822586,
 14.181176362883164,
 -41.22124451150056,
 19.41525454251834,
 39.316618656704335]

In [8]:
import math

def calcul_neurone(j, e, poids):
    ### START CODE HERE ###
    # The combination function is just the scalar product 
    # between e and the line j of the poids matrix
    resultat_j = sum([a * b for (a, b) in zip(e, [l[j] for l in poids])])
    ### END CODE HERE ###
    return int(resultat_j>0)



**Test** :

In [9]:
print(calcul_neurone(3, d[9][0], poids))
print(calcul_neurone(9, d[9][0], poids))

0
1


**Sortie attendue** :

    -123.05645010986458
    91.83223143508798

**Exercice** : Créer une fonction nommée *calcul_reseau* qui calcule toutes les sorties *j* d'un réseau de neurones à partir d'une entrée *e* représentant un chiffre et de la matrice *poids*.

In [10]:
def calcul_reseau(e, poids):
    ### START CODE HERE ###
    resultats = [calcul_neurone(i, e, poids) for i in range(10)]
    ### END CODE HERE ###
    return resultats


**Test** :

In [11]:
calcul_reseau(d[9][0], poids)

[1, 0, 1, 0, 0, 1, 1, 1, 0, 1]

**Sortie attendue** :

    [43.86749469120408,
     -253.71476511397273,
     107.07214011201678,
     -123.05645010986458,
     -122.1205211573054,
     14.24202347465732,
     17.195101727623452,
     47.77513612278679,
     -59.23046233947509,
     91.83223143508798]

## Apprentissage
L'apprentissage consiste à ajuster les poids de manière à ce que les sorties du réseau de neurones (calculées par *calcul_reseau*) correspondent bien aux chiffres fournis en entrée du réseau. Cet apprentissage est réalisé itérativement. A chaque itération on va faire progresser les poids, en confrontant les représentations des chiffres du dictionnaire *d* aux sorties calculées. Le nombre d'itérations est ici fixé. De manière imagée, vous aurez trois fonctions à développer, correspondant aux boucles suivantes :

boucle sur les itérations (*apprendre*)
-> boucle sur les entrees du dictionnaire (*apprendre_reseau*)
-> boucle sur les neurones (*apprendre_neurone*)

**Exercice** : Créer une fonction nommée *apprendre_neurone* qui permet à partir d'une entrée *e* représentant un chiffre d'ajuster la matrice *poids* de telle manière que la sortie calculée se rapproche de *1* si le chiffre *e* représente *j* ou de *0* sinon. L'argument *sortie_attendue* contient *1* si le chiffre *e* représente effectivement *j*, *0* sinon.

Cet apprentissage se base sur la formule suivante que nous vous fournissons, qui ajuste le poids du neurone situé entre la i-ième entrée et la j-ième sortie :

    poids[i][j] = poids[i][j] + (valeur_desiree - valeur_calculee) * e[i] * h

La variable *h* est un paramètre d'apprentissage, positif, que nous vous proposons de fixer à *5* et que vous pouvez faire varier.

Optimiser la matrice *poids* de telle façon que les sorties calculées la fonction *calcul_reseau* correspondent aux chiffres du dictionnaire.

In [12]:
def apprendre_neurone(e, poids, j, sortie_attendue):
    ### START CODE HERE ###
    h = 5
    # Calcul de la sortie du neurone j, avec l'entree e et la matrice de poids.
    valeur_calculee = calcul_neurone(j, e, poids)
    # Boucle sur les composantes de e et ajustement des poids correspondant avec la formule donnee.
    for i in range(len(poids)): # A MODIFIER
        # Calcul de la valeur desiree en fonction de sortie attendue.
        valeur_desiree = 0
        if sortie_attendue == j:
            valeur_desiree = 1
        # Actualisation du poids[i][j] (cf. formule precedente).
        poids[i][j] = poids[i][j] + (valeur_desiree - valeur_calculee) * e[i] * h
    ### END CODE HERE ###
    return poids


**Exercice** : Créer une fonction nommée *apprendre_reseau* qui permet, à partir d'une entrée *e* représentant un chiffre, d'ajuster la matrice *poids* en invoquant *apprendre_neurone* sur les différents neurones *j* du réseau.

In [13]:
def apprendre_reseau(e, poids, sortie_attendue):
    ### START CODE HERE ###
    for j in range(10):
        apprendre_neurone(e, poids, j, sortie_attendue)
    ### END CODE HERE ###


**Exercice** : Créer une fonction nommée *apprendre* qui permet, en bouclant sur les données, d'ajuster la matrice *poids* en invoquant *apprendre_reseau* sur les différentes représentations des chiffres.

In [14]:
def apprendre(d, poids):
    ### START CODE HERE ###
    for k in d:
        for e in d[k]:
            apprendre_reseau(e, poids, k)
    ### END CODE HERE ###


## Test du réseau de neurones
Vous disposez à présent des fonctions permettant d'entrainer un réseau de neurones à reconnaître des chiffres à partir d'images simples et de prédire une sortie associée à une entrée au clavier.

**Exercice** : Compléter le squelette du programme suivant qui :
- structure les données ;
- créée les poids ;
- créée le réseau de neurones ;
- permet au réseau de neurones d'apprendre à partir des données structurées ;
- permet à l'utilisateur de saisir les entrées à tester.

In [15]:
def main():
    ### START CODE HERE ###
    # Structuration des données
    d = structuration_donnees()
    # Création des poids
    poids = init_poids(30, 10)
    # Apprentissage
    nb_rounds = 100
    for i in range(nb_rounds):
        apprendre(d, poids)
    # Saisie de l'entree
    matrix_name = input("Entrez le nom de la matrice :")
    matrix = eval(matrix_name)
    m = flatten_matrice(matrix)
    s = "".join([str(i) for i in m])
    print("La matrice applatie est : {}".format(s))
    print("La matrice imprimée est : ")
    print(m)
    print_matrice(m)
    
    # Test de la conformité de la saisie (30 symboles '0' ou '1' séparés par une virgule)
    pass
    # Transformation de la saisie en tableau
    #input = None
    # Calcule toutes les sorties du réseau de neurones à partir de l'entrée input
    e = [int(i) for i in s]
    # print(poids)
    output = calcul_reseau(e, poids)
    # Affichage des sorties
    print(output)
    ### END CODE HERE ###



**Test** :

In [16]:
main()

Entrez le nom de la matrice :nb9a
La matrice applatie est : 111111000111111000010000111111
La matrice imprimée est : 
[1, 1, 1, 1, 1, 1, 0, 0, 0, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1]
#####
#   #
#####
    #
    #
#####
[0, 0, 0, 0, 0, 0, 0, 0, 0, 1]


In [17]:
d = structuration_donnees()
# Création des poids
poids = init_poids(30, 10)

def learn(nb_rounds, d, poids):
    for i in range(nb_rounds):
        apprendre(d, poids)

def identify(string_matrix):
    e = [int(i) for i in string_matrix]
    # print(poids)
    output = calcul_reseau(e, poids)
    # Affichage des sorties
    print(output)


In [18]:
learn(100, d, poids)

In [23]:
matrix_name = input("Entrez le nom de la matrice :")
matrix = eval(matrix_name)
m = flatten_matrice(matrix)
s = "".join([str(i) for i in m])
print("La matrice applatie est : {}".format(s))
print("La matrice imprimée est : ")
print_matrice(m)
s_noise = input("Entrez la matrice applatie avec du bruit :")
print("La matrice bruitée est :")
print_matrice([int(i) for i in s_noise])

identify(s_noise)

Entrez le nom de la matrice :nb2a
La matrice applatie est : 111110000111111100001000011111
La matrice imprimée est : 
#####
    #
#####
#    
#    
#####
Entrez la matrice applatie avec du bruit :110110000111111100001000011111
La matrice bruitée est :
## ##
    #
#####
#    
#    
#####
[0, 0, 1, 1, 0, 0, 0, 0, 1, 0]


In [20]:
s_noise = input("Entrez la matrice applatie avec du bruit :")
print("La matrice bruitée est :")
print_matrice([int(i) for i in s_noise])

identify(s_noise)

Entrez la matrice applatie avec du bruit :001000010000100001000010000111
La matrice bruitée est :
  #  
  #  
  #  
  #  
  #  
  ###
[0, 1, 0, 0, 1, 0, 0, 0, 0, 0]


## Optionnel : export jupyter